actions/tamper_action.go

```diff
@@ -86,60 +89,637 @@ func ParseTamperAction(s *scanner.Scanner) (Action, error) {
+			valueGen:     gen,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("invalid tamper rule: %q is not a valid tamper mode for IPv4", ta.Mode)
+}
+
+func (a *IPv4TamperAction) Apply(packet gopacket.Packet) ([]gopacket.Packet, error) {
+	ip := packet.Layer(layers.LayerTypeIPv4).(*layers.IPv4)
+	if ip == nil {
+		return nil, fmt.Errorf("packet does not have a IPv4 layer")
+	}
+
+	tamperIPv4(ip, a.field, a.valueGen)
+
+	return a.Action.Apply(packet)
+}
+
+// tamperIPv4 modifies the given IP field using the given value generator.
+func tamperIPv4(ip *layers.IPv4, field IPv4Field, valueGen tamperValueGen) error {
+	switch field {
+	case IPv4FieldSrcIP:
+		ip.SrcIP = valueGen.bytes(4)
+	case IPv4FieldDstIP:
+		ip.DstIP = valueGen.bytes(4)
+	case IPv4FieldVersion:
+		ip.Version = uint8(valueGen.uint(8))
+	case IPv4FieldIHL:
+		ip.IHL = uint8(valueGen.uint(8))
+	case IPv4FieldTOS:
+		ip.TOS = uint8(valueGen.uint(8))
+	case IPv4FieldLength:
+		ip.Length = uint16(valueGen.uint(16))
+	case IPv4FieldFlags:
+		// TODO: maybe implement this?
+	case IPv4FieldFragOffset:
+		ip.FragOffset = uint16(valueGen.uint(16))
+	case IPv4FieldTTL:
+		ip.TTL = uint8(valueGen.uint(8))
+	case IPv4FieldProtocol:
+		ip.Protocol = layers.IPProtocol(valueGen.uint(8))
+	case IPv4FieldChecksum:
+		ip.Checksum = uint16(valueGen.uint(16))
+	case IPv4Load:
+		ip.Payload = valueGen.bytes(1480)
 	}
 
-	if _, err = s.Expect(","); err == nil {
-		if !s.FindToken(")", true) {
-			return nil, errors.New("tamper rules can only have one action")
+	// let gopacket handle converting modified packet into []byte again, it's just easier
+	// again copy the bytes back into the packet header
+	sb := gopacket.NewSerializeBuffer()
+	ip.SerializeTo(sb, gopacket.SerializeOptions{})
+	ip.Contents = make([]byte, len(sb.Bytes()))
+	copy(ip.Contents, sb.Bytes())
+
+	// do we update length and checksum???
+	// it doesn't look like the geneva team does when they tamper with the IP layer
+
+	return nil
+}
+
+func updateIPv4LengthAndChksum(ip *layers.IPv4) {
+	length := len(ip.Contents) + len(ip.Payload)
+	ip.Length = uint16(length)
+	binary.BigEndian.PutUint16(ip.Contents[2:4], ip.Length)
+	buf := make([]byte, length)
+	copy(buf, ip.Contents)
+	copy(buf[len(ip.Contents):], ip.Payload)
+	chksum := checksum(buf)
+	ip.Checksum = chksum
+	binary.BigEndian.PutUint16(ip.Contents[10:12], chksum)
+}
+
+// copied directly from gopacket/layers/ip4.go because they didn't export one. for whatever some reason..
+func checksum(bytes []byte) uint16 {
+	// Clear checksum bytes
+	bytes[10] = 0
+	bytes[11] = 0
+
+	// Compute checksum
+	var csum uint32
+	for i := 0; i < len(bytes); i += 2 {
+		csum += uint32(bytes[i]) << 8
+		csum += uint32(bytes[i+1])
+	}
+	for {
+		// Break when sum is less or equals to 0xFFFF
+		if csum <= 65535 {
+			break
 		}
+		// Add carry to the sum
+		csum = (csum >> 16) + uint32(uint16(csum))
+	}
+	// Flip all the bits
+	return ^uint16(csum)
+}
+
+// Should we implement tamper for UDP??
```

Comment:
We probably should. _We_ don't use UDP for our own protocols, but as this is meant to be a general-purpose Go re-implementation of the original Python library, we should be as complete as we can.

Feel free to log a ticket for implementing the UDP tamper bits and we can say we'll do it later.

Reply:
My two cents:

- Since we have no need for UDP support ourselves, I say we leave it as a TODO.
- Since we're unlikely to develop a need for UDP support, I think a ticket is just going to gather dust until it's eventually closed by stale-bot.
- With this being an open source project, it should be clear to consumers of the library that UDP support is missing.

Given all of the above, I would:
- Add a disclaimer to the README indicating that UDP support is missing.  It would be nice to include some brief information indicating how an interested contributor could go about adding UDP support.
- Add one or more disclaimers in the Go documentation - wherever a caller might expect to find UDP support.
- (Optionally) have the tamper action notice when it's given a UDP packet and produce a meaningful error like "UDP tamper support is not implemented".

Reply:
Yeah, works for me.